topics/products/phpstorm/php_open_api_php_type_providers.md

```diff
@@ -0,0 +1,303 @@
+[//]: # (title: PHP Type Providers)
+
+<!-- Copyright 2000-2022 JetBrains s.r.o. and other contributors. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file. -->
+
+Type inference in PhpStorm is built on top of type providers, each of which is responsible for inferring the types of
+specific PSI elements.
+For example, `com.jetbrains.php.lang.psi.resolve.types.PhpArrayAccessTP` is responsible for inferring the types of expressions like `$arr[10]`.
+There are dozens of such providers, and they all work one after another to provide type information when needed.
+
+All providers inherit from `com.jetbrains.php.lang.psi.resolve.types.PhpTypeProvider4`, which is registered in the `com.jetbrains.php.typeProvider4` extension point.
+
+## Types in PhpStorm
+
+The first stage of type inference in PhpStorm takes place at the indexing stage, at this moment PhpStorm don’t
```

Comment:
To make this fully correct I'd also specify that during indexing it's possible to use content of local current file to obtain such information. For some cases it can represent effectively complete information for the project, for instance declaration and usages of private method in the current file.

Reply:
Fixed.